docs/docs/theory_2.md

```diff
@@ -48,6 +48,111 @@ This way we can see that the equations are completely decoupled. Coefficients on
 
 > **For developers:** time evolution of a system is handled by the `.evolve()` method of the `Hamiltonian` class.
 
+#### Celio's Method
+
+MuSpinSim can also make use of an approximation to speedup calculations and reduce memory usage in certain cases using [Celio's method](https://www.doi.org/10.1103/PhysRevLett.56.2720). To do this we first split up the Hamiltonian into contributions from each interaction.
+
+$$
+H = \sum_{i}^{N} H_i
+$$
+
+Then referring back to the earlier result
+
+$$
+\rho(t) = e^{-\frac{i}{\hbar}Ht}\rho(0)e^{\frac{i}{\hbar}Ht}.
+$$
+
+We expand using the Suzuki–Trotter formula
+
+$$
+e^{H_1 + H_2} = \lim_{k\rightarrow\infty}{\left[e^{\frac{H_1}{k}}e^{\frac{H_2}{k}}\right]^k}
+$$
+
+To obtain
+
+$$
+e^{-\frac{i}{\hbar}Ht} = \lim_{k\rightarrow\infty}{\left[\prod_{i}^{N}e^{-\frac{i}{k\hbar}H_it}\right]^k}
+$$
+
+This allows us to compute the evolution operator while avoiding the diagonalisation of the Hamiltonian. In reality this formula is a simplification as each $H_i$ acts in a smaller subspace of dimension determined by the spins involved in the interaction it describes. As a result, in computing this product in terms of matrices, we must also do the kronecker product with identity matrices that match the dimensions of the other particles in the system. We also use swap gates to ensure the order of these kronecker products is preserved.
+
+As an example, taking system of a muon and two electrons (labelled 1, 2 and 3 respectively) with a single dipolar interaction defined between the muon and second electron we compute
+
+$$
+e^{-\frac{i}{\hbar}Ht} = \lim_{k\rightarrow\infty}{\left[\text{SWAP}_{32} \left( \mathbb{1}_2 \otimes e^{-\frac{i}{k\hbar}H_{13}t}\right)\right]^k}
+$$
+
+Where $H_{12}$ is the contribution from the dipolar interaction and $\mathbb{1}_2$ is the identity matrix of size $2I + 1 = 2$ (For the first electron). $\text{SWAP}_{32}$ is a swap gate that has the effect of reversing the kronecker products into the correct order and is required since $H_{13}$ is formed in a subspace with only particles 1 and 3 whereas it should be computed for the system with particles 1, 2 and 3 in that order.
```

Comment:
So the role of the `SWAP` is the bit that I struggled with. I'm going to write what I undetstand we're doing here (which might be partially or completely wrong):

We can't make assumptions about how many hamiltonions we have, or which spins they apply to, and each one is `formed in a subspace` and so won't necessarily affect all spins in the system, or in the order that they are defined in the state vector `V`. We need the hamiltonions (`M` in the C++) to stay small so they don't blow up our memory, so instead of applying the `SWAP` gate to `M` (calculating an outer matrix product) or applying it directly to `V`, every time we perform an elementwise multiplication between `M` and `V` we choose the entry from `V` such that it has the same outcome as using the `SWAP`. In practice this means a different set of `indices` for each hamiltonian, as it will act on different spins/orders of spins. We only need to do this for the evolution and not the measure because when we measure we use the muon's spin operator, and the system will always be have the muon spin first (we even raise an error if it isn't).

If that's (mostly) right, then I suppose the current documentation is good enough as I was able to understand what's going on by reading it. Some areas that were confusing to me though:

- We refer to "reversing" the order of kronecker products here and in the docstrings of the C++ functions. Upon first reading I understood this to mean our `indices` would be (for example) `(0, 1, 2, 3, ...) -> (..., 3, 2, 1, 0)`, but in principle couldn't the "re-ordering" of `indices` be anything? E.g. `(0, 3, 1, 2, ...)`? Or does "reversing" mean something different in this context?
- This may not be of interest to users, but something to directly relate the role of `SWAP` to how it's implemented in the code (i.e. `indices` for each operator). Possibly in the same style as the other `**For developers:**` asides, or maybe be ammending the code comments/docstrings to relate it to the idea of `SWAP`s?
- Again either in the code or the documentation here, explicitly state some of other points I assumed (if the assumptions are correct). Why we don't need to do it for the `measure`, the fact that if the system had another hamiltonion we would need a distinct `SWAP`/`indices` ordering for each of them.

I appreciate this is quite a long, vague comment. If anything I've understood is wrong, or would be so obvious to someone who's actively involved in the domain, then do let me know - equally if you have any preferences for the format and location of any potential changes to the documentation.

Reply:
@patrick-austin Yes that sounds right to me. Effectively if we don’t use Celio’s each interaction’s contribution to the Hamiltonian is the result of Kronecker products of the relevant spin matrices for the spins that are interacting and otherwise an identity matrix of the same size but always in the order they are defined in the system. For Celio’s we ignore the identity matrices until later so if we were doing something like `A x I x B`, we would now be doing `A x B x I`. The swap operator makes this switch by swapping the relevant rows and columns of the combined matrix. For the python version we still compute A x B x I and then swap the rows/columns using QuTiP, but the C++ variant goes further, ignoring the identity matrix, I, entirely. (For the example we tested we only store 16x16 matrices instead of doing a Kronecker product with an identity matrix of size 8^4 which is bad for memory). This method is extremely slow in python due to array access & loop speeds. The indices could indeed be anything - it will only have the reversing effect if chosen appropriately – I should make that clearer.

The operator used in the measure method works in the same way. It is composed of a linear combination of its pauli matrices followed by Kronecker products with the identities for each other spins in the system. These identity matrix Kronecker products just form a larger identity matrix. By forcing the muon to always be first in the system the swapping is not needed since the order of the rest of the spins just reorders the identity matrices which has no effect.

I will add it to the C++ comments since its quite specific to where its needed.

Reply:
I have just amended the comments, hopefully that's clearer.